dipy/align/imaffine.py

```diff
@@ -0,0 +1,777 @@
+import numpy as np
+import numpy.linalg as npl
+import scipy.ndimage as ndimage
+from dipy.align import floating
+import dipy.align.vector_fields as vf
+from dipy.align.mattes import MattesBase, sample_domain_regular
+from dipy.core.optimize import Optimizer
+from dipy.align.imwarp import (get_direction_and_spacings, ScaleSpace)
+from dipy.align.scalespace import IsotropicScaleSpace
+from dipy.core.optimize import SCIPY_LESS_0_12
+
+
+class MattesMIMetric(MattesBase):
+    def __init__(self, nbins=32, sampling_proportion=None):
+        r""" Initializes an instance of the Mattes MI metric
+
+        This class implements the methods required by Optimizer to drive the
+        registration process by making calls to the low level methods defined
+        in MattesBase.
+
+        Parameters
+        ----------
+        nbins : int, optional
+            the number of bins to be used for computing the intensity
+            histograms. The default is 32.
+        sampling_proportion : int in (0,100], optional
+            the percentage of voxels to be used for estimating the (joint and
+            marginal) intensity histograms. If None, dense sampling is used
+            (see notes regarding the difference between sparse and dense
+            sampling). The default is None.
+
+        Notes
+        -----
+        Since we use linear interpolation, images are not, in general,
+        differentiable at exact voxel coordinates, but they are differentiable
+        between voxel coordinates. When using sparse sampling, selected voxels
+        are slightly moved by adding a small random displacement within one
+        voxel to prevent sampling points to be located exactly at voxel
+        coordinates. When using dense sampling, this random displacement is
+        not applied.
+        """
+        super(MattesMIMetric, self).__init__(nbins)
+        self.sampling_proportion = sampling_proportion
+
+    def setup(self, transform, static, moving, static_grid2space=None,
+              moving_grid2space=None, prealign=None):
+        r""" Prepares the metric to compute intensity densities and gradients
+
+        The histograms will be setup to compute probability densities of
+        intensities within the minimum and maximum values of `static` and
+        `moving`
+
+        Parameters
+        ----------
+        transform: instance of Transform
+            the transformation with respect to whose parameters the gradient
+            must be computed
+        static : array, shape (S, R, C) or (R, C)
+            static image
+        moving : array, shape (S', R', C') or (R', C')
```

Comment:
That's right!, the dimensions of the input images don't need to be the same. I added a clarifying note.